// @code-bucket/board-game-geek/dist/models/bgg-link.model.js
export var BggLinkType;
(function (BggLinkType) {
    BggLinkType["boardGameCategory"] = "boardgamecategory";
    BggLinkType["boardGameMechanic"] = "boardgamemechanic";
    BggLinkType["boardGameFamily"] = "boardgamefamily";
    BggLinkType["boardGameExpansion"] = "boardgameexpansion";
    BggLinkType["boardGameImplementation"] = "boardgameimplementation";
    BggLinkType["boardGameDesigner"] = "boardgamedesigner";
    BggLinkType["boardGameArtist"] = "boardgameartist";
    BggLinkType["boardGamePublisher"] = "boardgamepublisher";
    BggLinkType["boardGameVersion"] = "boardgameversion";
    BggLinkType["language"] = "language";
    BggLinkType["boardGameAccessory"] = "boardgameaccessory";
})(BggLinkType || (BggLinkType = {}));
var BggLink = /** @class */ (function () {
    function BggLink(data) {
        var _a;
        this.id = Number.parseInt(data._attributes.id);
        this.type = data._attributes.type;
        this.value = data._attributes.value;
        this.inbound = ((_a = data._attributes) === null || _a === void 0 ? void 0 : _a.inbound) === 'true';
    }
    return BggLink;
}());
export { BggLink };
